### ISSUE TITLE
[NEXT-471] Calling cookies.set twice sets only one value, in API route with Edge runtime

### ISSUE THREAD

**Issue by eric-burel** [vercel/next.js#38302]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

Not applicable, see reproduction here: https://codesandbox.io/s/condescending-torvalds-e3il3k?file=/package.json

### Describe the Bug

Calling "cookies.set" twice on an API route with edge handler runtime will set only one cookie.

### Expected Behavior

You'll see one "Set-Cookie" with 2 values, however it should be instead 2 "Set-Cookie" with only one value.

Set-Cookie can be repeated as many time as needed in an HTTP response to set multiple cookies. Edge middlewares do not suffer from this issue, it affects only API routes.

### Link to reproduction

https://codesandbox.io/s/condescending-torvalds-e3il3k

### To Reproduce

https://codesandbox.io/s/condescending-torvalds-e3il3k?file=/pages/index.tsx
Open index, check the Network tab for the "cookie" request. 

<sub>[NEXT-471](https://linear.app/vercel/issue/NEXT-471/calling-cookiesset-twice-sets-only-one-value-in-api-route-with-edge)</sub>

**Comment by controversial** [comment]
Note this issue still exists even when returning a `Response`, even when avoiding the `NextResponse` cookies API entirely:

Starting from `npx create-next-app -e reproduction-template`, I added only one file, `pages/api/test.js`:
```js
export default function handler() {
  const res = new Response();
  res.headers.append('Set-Cookie', 'a=b');
  res.headers.append('Set-Cookie', 'c=d');
  return res;
}

export const config = {
  runtime: 'experimental-edge',
};
```

Now, running `npm run dev` and `curl -v http://localhost:3000/api/test`:
```
*   Trying 127.0.0.1:3000...
* Connected to localhost (127.0.0.1) port 3000 (#0)
> GET /api/test HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.79.1
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< set-cookie: a=b, c=d
< Date: Fri, 15 Jul 2022 18:47:49 GMT
< Connection: keep-alive
< Keep-Alive: timeout=5
< Transfer-Encoding: chunked
< 
* Connection #0 to host localhost left intact
```
The expected behavior, when using `append` with `set-cookie`, is that multiple set-cookie headers are included in the response [[ref]](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#:~:text=To%20send%20multiple%20cookies%2C%20multiple%20Set%2DCookie%20headers%20should%20be%20sent%20in%20the%20same%20response.).

**Comment by controversial** [comment]
I’ve verified that this bug exists on `next@canary` so the “please verify canary” label should be removed.

**Comment by controversial** [comment]
Also worth noting that there’s a test [[a]](https://github.com/vercel/next.js/blob/607ff2b3225ea9fab04b4fca07e4a609b05e8038/test/e2e/middleware-responses/app/middleware.js#L42-L45), [[b]](https://github.com/vercel/next.js/blob/607ff2b3225ea9fab04b4fca07e4a609b05e8038/test/e2e/middleware-responses/test/index.test.ts#L30-L33) for this edge case for middleware. A similar test case is missing for edge functions

**Comment by eric-burel** [comment]
> Also worth noting that there’s a test [[a]](https://github.com/vercel/next.js/blob/607ff2b3225ea9fab04b4fca07e4a609b05e8038/test/e2e/middleware-responses/app/middleware.js#L42-L45), [[b]](https://github.com/vercel/next.js/blob/607ff2b3225ea9fab04b4fca07e4a609b05e8038/test/e2e/middleware-responses/test/index.test.ts#L30-L33) for this edge case for middleware. A similar test case is missing for edge functions

Thanks for the tip, I've started to write a test case: https://github.com/vercel/next.js/issues/38302#issuecomment-1185838007
I am stuck on running the test locally right now ^^ but I hope it will help getting some progress on the issue

**Comment by eric-burel** [comment]
Btw this seems vaguely related to https://github.com/whatwg/fetch/pull/1346, it discusses the handling of list of set-header values in fetch.

**Comment by controversial** [comment]
This issue still exists in Next.js 13.1, where [edge API routes have been marked “stable.”](https://nextjs.org/blog/next-13-1#a-light-nodejs-runtime-for-the-edge-now-stable-for-api-routes)

**Comment by alvarlagerlof** [comment]
This issue promoted me to store a stringified json object in a cookie. Not a great experience.

**Comment by sannajammeh** [comment]
Still a blocking problem here for introducing edge authentication... Not sure why the API is marked as stable when this issue persists into 2023.

**Comment by balazsorban44** [comment]
[https://github.com/vercel/edge-runtime/pull/255](https://github.com/vercel/edge-runtime/pull/255) needs to land first, then we can update Next.js as well

**Comment by controversial** [comment]
Glad to see [`@edge-runtime/cookies@3.0.3`](https://github.com/vercel/edge-runtime/releases/tag/%40edge-runtime%2Fcookies%403.0.3) land @balazsorban44! What still needs to be done in #42736 to incorporate these changes into Next.js?

**Comment by balazsorban44** [comment]
Getting the PR checks to pass. :grin: Working on it today.

**Comment by y0zong** [comment]
this seems still a issue even I upgrade to v13.2.1(which a fix has merged in v13.2.0), what did I miss?

**Comment by controversial** [comment]
Confirmed ^ On `next@13.2.2-canary.1`, the following still only produces only one Set-Cookie header:

```js
export default function handler() {
  const res = new Response('hey');
  res.headers.append('Set-Cookie', 'a=b');
  res.headers.append('Set-Cookie', 'c=d');
  return res;
}

export const config = {
  runtime: 'edge',
};
```

<img width="386" alt="Screenshot 2023-02-26 at 1 18 46 AM" src="https://user-images.githubusercontent.com/10377391/221395522-dfaf8e09-5649-41b3-8044-91490780719c.png">

@balazsorban44 should this issue be reopened?

**Comment by balazsorban44** [comment]
@controversial could you open a new one with the above reproduction? :pray: would be easier to track.

**Comment by y0zong** [comment]
> this seems still a issue even I upgrade to v13.2.1(which a fix has merged in v13.2.0), what did I miss?

I found that not only `set` cookie twice not work, seems any operation on cookie just *run* once, that's say
``` js
res.cookies.delete(oldkey)    // works
res.cookies.set(newkey, val)  // ignored

res.cookies.set(somekey)      // works
res.cookies.set(somekey, val) // ignored

res.cookies.set(key1)         // works
res.cookies.delete(key2, val) // ignored
```
not only `cookies` also `res.headers.set("set-cookie")` are same behavior, maybe It's not `@edge-runtime/cookies`'s problem I think?

**Comment by balazsorban44** [comment]
By the way, it's a different (but related) issue, since you were not using `NextResponse` as the OP did.

**Comment by controversial** [comment]
@balazsorban44 Happy to open a new issue; will do that in a moment!

I’m not sure this issue is solved either, though; This example uses the NextResponse API and is almost identical to the reproduction in the original issue:

<img width="599" alt="Screenshot 2023-02-28 at 3 37 35 PM" src="https://user-images.githubusercontent.com/10377391/221973940-429c6f8f-6b3f-461a-aecd-9146e883f3a8.png">

Only a single Set-Cookie header is returned.

Is there a different test case associated with this issue that #42736 fixed?

**Comment by bpossolo** [comment]
@balazsorban44 @controversial 
im seeing the same problem on NextJS `13.2.4`
using `NextResponse.cookies.set()` to set multiple cookies yields a single `Set-Cookie` http response header...
```
const response = NextResponse.redirect(location, 303);
response.cookies.set({
  name: ACCESS_TOKEN,
  value: result.accessToken,
  path: '/',
  secure,
  expires: result.expiresOn ? result.expiresOn : undefined,
});
response.cookies.set({
  name: ACCOUNT_ID,
  value: result.account.homeAccountId,
  path: '/',
  secure,
  expires: expiration('years', 1),
});
return response;
```
```
set-cookie: access-token=mytokenvalue; Path=/; Expires=Mon, 27 Mar 2023 04:38:03 GMT, account-id=1c12dd32-abcd-4e0a-9d21-5b480b806b06-b2c_1a_passwordless-authentication.abcdedf-1136-40a6-a376-123456789; Path=/; Expires=Tue, 26 Mar 2024 04:44:43 GMT
```

from [the spec](https://www.rfc-editor.org/rfc/rfc6265#section-3):
> Origin servers SHOULD NOT fold multiple Set-Cookie header fields into a single header field.  The usual mechanism for folding HTTP headers fields (i.e., as defined in [[RFC2616](https://www.rfc-editor.org/rfc/rfc2616)]) might change the semantics of the Set-Cookie header field because the %x2C (",") character is used by Set-Cookie in a way that conflicts with such folding.
comma is used in the Expires attribute value

**Comment by controversial** [comment]
I think this issue should be reopened

**Comment by petersng** [comment]
anyone know of a workaround for this issue?  it's currently a blocker for me.  thanks for any insights.

**Comment by eric-burel** [comment]
Note that I don't have the option to reopen as issue creator, you might want to open a separate issue to get attention on this issue. 
Also, I am pretty sure it can be a nice first contribution for whoever have time to check it out, it's (probably) not difficult since a similar logic exists in other part of the code like middlewares, node.js API routes, but requires digging the codebase.
